server/src/main/java/org/elasticsearch/action/support/replication/ReplicationRequest.java

```diff
@@ -197,6 +210,23 @@ public void writeTo(StreamOutput out) throws IOException {
         out.writeVLong(routedBasedOnClusterVersion);
     }
 
+    /**
+     * Thin serialization that does not write {@link #shardId} and will only write {@link #index} if it is different from the index name in
+     * {@link #shardId}.
+     */
+    public void writeThin(StreamOutput out) throws IOException {
+        super.writeTo(out);
+        waitForActiveShards.writeTo(out);
+        out.writeTimeValue(timeout);
+        if (shardId != null && index.equals(shardId.getIndexName())) {
```

Comment:
This `null` check is kind of annoying. In many cases we logically wouldn't need it (technically) because we are serializing a request instance that doesn't have the shard id set yet as part of a bulk request (but is known) on the coordinating node (when first sorting the various bulk request parts per shard). We can make a follow-up improvement here to make use of the known `ShardId` in these cases so that we don't needlessly serialize the index name if it isn't an alias.